### DISCUSSION TITLE
programmatically-changing-app-icon/

### DISCUSSION THREAD

**Discussion by giscus** [JakeSteam/blog-programming#25] [Comments]
# programmatically-changing-app-icon/

In Android, your app’s icon is a crucial part of your project. It’s what grabs the user’s attention first, identifies your app amongst the sea of competitors, and is the most important piece of visual branding. However, usually this icon is defined in your manifest, then never changes except through update.

https://blog.jakelee.co.uk/programmatically-changing-app-icon/

**Comment by csantarin** [comment]
Hey there, Jake. Thanks for writing this up. It's amazingly helpful.

What if the original `<activity>` tag contains more than one `<intent-filter>`?

Example:

```xml
<activity android:name=".MainActivity">
    <intent-filter>
        <category android:name="android.intent.category.LAUNCHER"/>
    </intent-filter>

    <!-- deep link and push notification intent filter -->
    <intent-filter android:autoVerify="true">
        <action android:name="android.intent.action.VIEW" />
        <category android:name="android.intent.category.DEFAULT" />
        <category android:name="android.intent.category.BROWSABLE" />
        <data android:scheme="https" android:host="deeplinkandpushnotification.provider.com" />
    </intent-filter>

    <!-- firebase cloud messaging intent filter -->
    <intent-filter>
        <action android:name="fcm.ACTION.HELLO" />

        <category android:name="android.intent.category.DEFAULT" />
    </intent-filter>
</activity>
```

Where do they need to go? Is it essential that we clone those `<intent-filter>`s, too? And for each new `<activity-alias>`? Or is it fine to retain those `<intent-filter>`s back at the `<activity>` tag?

```xml

<activity android:name=".MainActivity">
    <intent-filter>
        <category android:name="android.intent.category.LAUNCHER"/>
    </intent-filter>
</activity>

<activity-alias
        android:icon="@mipmap/icon_red"
        ...>
    <!-- launcher intent filter -->
    <intent-filter>
        <action android:name="android.intent.action.MAIN" />
        <category android:name="android.intent.category.LAUNCHER" />
    </intent-filter>

    <!-- deep link and push notification intent filter -->
    <!-- firebase cloud messaging intent filter -->
</activity-alias>

<activity-alias
        android:icon="@mipmap/icon_green"
        ...>
    <!-- launcher intent filter -->
    <!-- deep link and push notification intent filter -->
    <!-- firebase cloud messaging intent filter -->
</activity-alias>

<activity-alias
        android:icon="@mipmap/icon_blue"
        ...>
    <!-- launcher intent filter -->
    <!-- deep link and push notification intent filter -->
    <!-- firebase cloud messaging intent filter -->
</activity-alias>
```

I've had folks at work run up to me saying that not having copies of each `<intent-filter>` per `<activity-alias>` is dangerous, but I want to refute that. It's particularly because having a sudden explosion of XML code in the manifest just to support more icons for 1:1 functionality is going to make maintenance very costly — every icon added would cost us! If there was a way to make this cheap, I'd be all in for it!

I did attempt to only add exactly the necessary `<intent-filter>` in this tutorial to make the app launchable from those icons without changing anything else. Deep links did work, so I assume there really is no problem.

**Reply by JakeSteam** [reply]
Hey, good question!

I believe you don't need to redeclare any of the other intent filters, since the `targetActivity` is still going to be your "Main" activity. This article is a few years old though, I'd definitely recommend giving it a try and finding out for certain.

**Reply by csantarin** [reply]
@JakeSteam Ah-ha! I <ins>**knew**</ins> this was going to be a bit of an unknown! 😂 

> since the targetActivity is still going to be your "Main" activity.

I think you hit the head on the head with this one. I added more detail to my use case about having to test deep links. It still works! So, I suppose I'm safe...

Also, huge thanks again for replying. That's a rare gem in my experience.

**Comment by csantarin** [comment]
> If you are trying to deploy an app that has had it’s alias changed since install, Android Studio will fail to launch it. It will need to be uninstalled or reset back to the default.

I found a way to work around this while in development.

You can simply edit the launch configuration of the app using **Run/Debug Configurations** (Android Studio > Run > Edit Configurations...) and the IDE will launch it again using the designated `<activity-alias>`, no problem.

At the Run/Debug Configurations, select the launch configuration of your project (I assume `app` in my examples):
- Launch: `Specified Activity`
- Activity: The full `<activity-alias>` name of the icon you last selected. No variables!
- ✅  **Skip Activity validation** to ignore the flaky IDE activity check

![image](https://github.com/JakeSteam/blog-programming/assets/67677362/3062429e-77c1-4324-a8bb-0c0b4772f831)

Setting it back is easy when you switch back to the default icon. Just do:
- Launch: `Default Activity`

I assume that your project is configured to ignore these changes from commits. Otherwise, I'd highly recommend it in order to not mess the original configuration up!

**Comment by akardas16** [comment]
Hi thanks for the explanation. When I change the icon, app will be killed by the system. I mean PackageManager.DONT_KILL_APP is not working. Is there a way to prevent that?

**Reply by rovedashcam** [reply]
Hi Thanks for  posting this code . 
but one thing i just want to know can we change app icon without restart app when we change icon from old to new or new old it always  restart app  can we prevent this ?
